nameko/dependencies.py

```diff
@@ -68,75 +69,71 @@ def kill(self):
         swallow the exception to allow the container kill to continue.
         """
 
-    def worker_setup(self, worker_ctx):
-        """ Called before a service worker executes a task. This method is
-        called for all DependencyProviders, not just the one that triggered
-        the worker spawn.
-
-        DependencyProviders should do any pre-processing here, raising
-        exceptions in the event of failure.
-
-        Example: ...
-
-        Args:
-            - worker_ctx: see
-                ``nameko.containers.ServiceContainer.spawn_worker``
-        """
-
-    def worker_teardown(self, worker_ctx):
-        """ Called after a service worker has executed a task. This method is
-        called for all DependencyProviders, not just the one that triggered
-        the worker spawn.
-
-        DependencyProviders should do any post-processing here, raising
-        exceptions in the event of failure.
-
-        Example: a database session provider may commit the session
-
-        Args:
-            - worker_ctx: see
-                ``nameko.containers.ServiceContainer.spawn_worker``
-        """
-
     def bind(self, name, container):
-        """ Bind this DependencyProvider instance to ``container`` using the
-        given ``name`` to identify the resource on the hosted service.
-
-        Called during ServiceContainer initialisation. The DependencyProvider
-        instance is created and then bound to the ServiceContainer instance
-        controlling its lifecyle.
+        """ Bind a clone of this extension to `container` with the given
+        `name`.
         """
-        self.name = name
-        self.container = container
-        self.bound = True
+        if self.__clone:  # TODO: prefer this over check in clone?
+            raise RuntimeError('Cloned extensions cannot be bound.')
+
+        # if this extension is shared and there's already a cloned instance,
+        # return that
+        if self.__shared:
+            shared_extensions.setdefault(container, {})
+            shared = shared_extensions[container].get(self.sharing_key)
+            if shared:
+                return shared
+
+        # clone this extension and associate it with the container
+        clone = self.clone()
+        clone.name = name
+        clone.container = container
+
+        # recursively bind nested extensions
+        for ext_name, ext in inspect.getmembers(self, is_extension):
+            setattr(clone, ext_name, ext.bind(ext_name, container))
+
+        # if this extension is shared, save the cloned instance
+        if self.__shared:
+            shared_extensions[container][self.sharing_key] = clone
+
+        return clone
+
+    def clone(self):
+        if self.__state is None:
+            raise RuntimeError('Undefined extension state. Did you forget to '
+                               'call super().__init__() in your subclass?')
+
+        if self.__clone:  # TODO: need this even with the check in bind above?
```

Comment:
It's possible we'll lose these. It is mostly me asserting that my implementation worked correctly, but you're not supposed to make clones of clones. It might work but it's not correct.

An alternative would be to delete the clone method during the clone process. That would ensure safety without the extra code, but might be more difficult to understand for any developer coming across it.